File: simplerpc/src/main/java/com/springmarker/simplerpc/core/server/ProxyServerCore.java
package com.springmarker.simplerpc.core.server;

import com.springmarker.simplerpc.pojo.RpcRequest;
import com.springmarker.simplerpc.pojo.RpcResponse;

import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;


/**
 * @author Springmarker
 * @date 2018/10/15 21:23
 */
public class ProxyServerCore {

    private Receiver receiver;

    public ProxyServerCore(Receiver receiver) {
        this.receiver = receiver;
    }

    public RpcResponse handleMethod(RpcRequest rpcRequest) {
        ArrayList<Object> paramList = rpcRequest.getParamList();
        return receiver.receive(rpcRequest.getMethodHashCode(), rpcRequest.getParamList());
    }

    public void handleAsyncMethod(RpcRequest rpcRequest, CompletableFuture future) {
        int methodHashCode = rpcRequest.getMethodHashCode();
        ArrayList<Object> paramList = rpcRequest.getParamList();
        receiver.receiveAsync(methodHashCode, paramList, future);
    }

}

File: simplerpc/src/main/kotlin/com/springmarker/simplerpc/core/server/ProxyServerCore.kt
package com.springmarker.simplerpc.core.server

import com.springmarker.simplerpc.pojo.ExchangeRequest
import com.springmarker.simplerpc.pojo.ServerConfig
import net.sf.cglib.proxy.MethodProxy
import java.lang.reflect.Method


/**
 * @author Springmarker
 * @date 2018/10/15 21:23
 */
class ProxyServerCore(
        private val receiver: ReceiverInterface,
        private val serverHandler: ServerHandler,
        private val serverConfig: ServerConfig
) {

    init {
        serverHandler.start(serverConfig, this)
    }

    fun handleMethod(baseExchangeRequest: ExchangeRequest): Any? {
        return receiver.receive(baseExchangeRequest.rpcRequest)
    }

    /**
     * 处理同步方法
     */
    private fun handleSyncRequest(obj: Any?, method: Method, args: Array<out Any>, proxy: MethodProxy?): Any? {
        this.receiver.receive(null)
        return null
    }

}